persistent/persistence.py

```diff
@@ -331,36 +361,81 @@ def __reduce__(self):
-                    self.__jar.setstate(self)
-                except:
-                    self.__flags = before
-                    raise
+        oga = _OGA
+        before = oga(self, '_Persistent__flags')
+        if before is None: # Only do this if we're a ghost
+            # Begin by marking up-to-date in case we bail early
+            _OSA(self, '_Persistent__flags', 0)
+            jar = oga(self, '_Persistent__jar')
+            if jar is None:
+                return
+            oid = oga(self, '_Persistent__oid')
+            if oid is None:
+                return
+
+            # If we're actually going to execute a set-state,
+            # mark as changed to prevent any recursive call
+            # (actually, our earlier check that we're a ghost should
+            # prevent this, but the C implementation sets it to changed
+            # while calling jar.setstate, and this is observable to clients).
+            # The main point of this is to prevent changes made during
+            # setstate from registering the object with the jar.
+            _OSA(self, '_Persistent__flags', CHANGED)
+            try:
+                jar.setstate(self)
+            except:
+                _OSA(self, '_Persistent__flags', before)
+                raise
+            else:
+                # If we succeed, no matter what the implementation
+                # of setstate did, mark ourself as up-to-date. The
+                # C implementation unconditionally does this.
+                _OSA(self, '_Persistent__flags', 0) # up-to-date
+
+    # In the C implementation, _p_invalidate winds up calling
+    # _p_deactivate. There are ZODB tests that depend on this;
+    # it's not documented but there may be code in the wild
+    # that does as well
 
     def _p_deactivate(self):
         """ See IPersistent.
         """
         if self.__flags is not None and not self.__flags:
-            self._p_invalidate()
+            self._p_invalidate_deactivate_helper()
 
     def _p_invalidate(self):
         """ See IPersistent.
         """
+        # If we think we have changes, we must pretend
+        # like we don't so that deactivate does its job
+        _OSA(self, '_Persistent__flags', 0)
+        self._p_deactivate()
+
+    def _p_invalidate_deactivate_helper(self):
         if self.__jar is not None:
             if self.__flags is not None:
-                self.__flags = None
+                _OSA(self, '_Persistent__flags', None)
             idict = getattr(self, '__dict__', None)
             if idict is not None:
                 idict.clear()
+            # Implementation detail: deactivating/invalidating
+            # updates the size of the cache (if we have one)
+            # by telling it this object no longer takes any bytes
+            # (-1 is a magic number to compensate for the implementation,
+            # which always adds one to the size given)
+            try:
+                cache = self.__jar._cache
```

Comment:
ISTM that using `_OGA()` everywhere for the dunder names is clearer:  otherwise, folks maintaining the module six months from now will puzzle over it, not recall the context, and maybe pick wrong.